core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -3,156 +3,406 @@
 package scalaz.zio
 
 import scala.collection.immutable.{ Queue => IQueue }
-
-import internal._
+import Queue.internal._
 
 /**
- * A `Queue` is a lightweight, asynchronous queue. This implementation is
- * naive, if functional, and could benefit from significant optimization.
- *
- * TODO:
- *
- * 1. Investigate using a faster option than `Queue`, because `Queue` has
- *    `O(n)` `length` method.
- * 2. Benchmark to see how slow this implementation is and if there are any
- *    easy ways to improve performance.
- */
-class Queue[A] private (capacity: Int, ref: Ref[State[A]]) extends Serializable {
+ * A `Queue` is a lightweight, asynchronous queue. This implementation is
+ * naive, if functional, and could benefit from significant optimization.
+ *
+ * TODO:
```

Comment:
These todos can be deleted. We know we need to optimize performance, it's the next major piece of work for `Queue`.